[Feat] YW2-25 공통 컴포넌트 -TextField 구현

Sangjun-man opened the pull request:
## 스크린샷
> /text-field  url에 접근하면 예시로 동작하는 컴포넌트를 확인하실수 있습니다.


https://github.com/YAPP-Github/22nd-Web-Team-2-Web/assets/66112027/1df61f10-7c82-4857-bc4e-5d08ad0d5bf7



https://github.com/YAPP-Github/22nd-Web-Team-2-Web/assets/66112027/48c58426-84f4-420b-b5f1-aad42ed0b504

https://github.com/YAPP-Github/22nd-Web-Team-2-Web/assets/66112027/ccdb5576-7045-45e9-9f67-f1f5ddf1a7c9




<br>

## Motivation

- #12 

### 프로젝트에서 값 입력에 사용되는 TextField 공통 컴포넌트를 구현했습니다
  TextField 컴포넌트의 TextFieldProps 타입 설명
  - name : TextField를 구분하는 이름,
  - type : input 타입을 설정, text 또는 password
  - label : 입력창 위쪽에 위치하는 label
  - message : 입력창 아래쪽에 위치하는 작성 안내 메시지
  - status : 외부에서 TextField 창의 상태를 설정해줄 때 사용, 인풋창의 상태를 바꿉니다.
  - validation : 입력값 유효성 검증 옵션, 현재 최대길이를 검증하는 "max"와 이메일을 검증하는 "email" 만 구현되어있습니다.
  - errorCallback : 콜백함수를 인자로 받으며, 입력창 내에서 에러 발생시 콜백함수의 첫번째 파라미터로 에러가 발생한 TextField의 name을 리턴합니다. {name:string} 형태로 리턴
  - defaultValue : 텍스트필드 생성시 기본적으로 입력되어있는 값
  - onChange, onBlur : react-hook-form과 연동을 위해 존재하는 props


### 외부에서 input창에 상태 전달하기
TextField의 상태는 5가지로 구성해두었습니다. 
```typescript
type TextFieldStatus =
  | 'default'
  | 'loading'
  | 'active'
  | 'success'
  | 'error';
```

TextField 컴포넌트에서는 props로 전달받는 status를 deps로 가지는 useEffect가 구현되어 있어서 props로 status를 전달해줄 시 인풋창의 상태를 바꿉니다. 이 기능을 통해 loading, success 상태 또한 지정해줄 수 있고, 추후 보호소 이름 중복 확인 api 요청시 loading, response 결과에 따라 success 또는 error 상태를 전달하는 방식으로 인풋창을 컨트롤 할 수 있습니다. status 전달과 함께 messge도 전달하는 경우 해당 메시지를 반영해서 보여줄 수 있도록 구현했습니다.

### react-hook-form 연동하기
react-hook-form 의 register 메서드를 스프레드 형식으로 TextField 컴포넌트에 추가해주시면, 인풋 data에 접근할 수 있습니다.
```jsx
 <TextField
          label="라벨"
          placeholder="플레이스홀더"
          message="전달 메시지"
          {...register('인풋5')} //react-form-hook과 함께 사용, ref 추가할 필요 없음
        />
```

### validation, error 관련
validation의 경우, yup 라이브러리를 활용해서 현재 최대값, 이메일 확인만 구현해 두었는데요 추후에 필요한 내용이 있으면 추가하겠습니다


현재 TextField 외부에 error를 전달하는 방식은 다음과 같습니다
TextField 내부에는 'status' 를 deps로 갖는 useEffect가 있습니다. 여기서 status가 error일 때만 props로 전달받은 errorCallback 함수의 첫번째 인자에 에러가 발생한 컴포넌트의 name을 넣어서 실행시킵니다. 현재는 이름에 대한 정보만 errorCallback에서 전달하는데, 추후에 다른 상태나 어떤 에러가 발생했는지 등에 대한 내용이 필요하면 추가하겠습니다.

<br>

## To Reviewers

- 어떤거든 조아요,, 피드백 주시면 너무 감사하겠습니다
- 제생각에는 textField 상태처리, style 관련, handler 쪽 코드가 조금 아쉬운것 같습니다

### Key Changes

- SVG 에셋 사용을 위해 SVGR 라이브러리를 설치했습니다.

Sangjun-man committed:
✨ feat: textField 기본 기능 구현

- 기능 구현 목록
  글자 카운트 기능 구현
  label, placeholder 구현
  글자 카운트 초과 시 색깔 변화 구현

Sangjun-man committed:
Merge branch 'develop' into feat/YW2-14@기본_컴포넌트_구현

Sangjun-man committed:
✨ feat: 글로벌 스타일 적용

Sangjun-man committed:
🔨 fix: textField ref 적용 로직 수정, 핸들러 로직 수정

Sangjun-man committed:
✨ feat: textField에 사용되는 state hooks로 분리

Sangjun-man committed:
🤖 refactor: useTextField 내부 state 형태, 로직 수정

Sangjun-man committed:
🤖 refactor: eventHandler내의 Ref null체크를 getInitialRef 유틸함수로 분리하여 처리

Sangjun-man committed:
💄 style: 주석 추가, console.log 제거

Sangjun-man committed:
🔨 fix: 최대값 validation 설정 안됐을 시 글자수 count하는 ref 적용되지 않던 에러 수정

Sangjun-man committed:
🔨 fix: textField valdation 로직 수정

Sangjun-man committed:
✨ feat: defaultValue 추가 기능 구현

Sangjun-man committed:
🤖 refactor: 인풋값 길이를 카운트하는 string을 생성하는 유틸함수 추가

Sangjun-man committed:
✨ feat: input 비밀번호 type 설정 기능 추가

Sangjun-man committed:
🚚 chore: svgr 패키지 추가

Sangjun-man committed:
🔨 fix: textField 컴포넌트 스타일 수정

Sangjun-man committed:
✨ feat: 입력값 상태에 맞게 remove버튼 랜더링하는 기능 추가

Sangjun-man committed:
🔨 fix: textField 기본 메시지 에러 제외 없도록 수정

Sangjun-man committed:
🚚 chore: textField 예시 페이지 생성

naro-Kim commented on src/components/TextField/TextField.tsx:
```diff
@@ -0,0 +1,247 @@
+          onFocus={handleFocus}
+          ref={node => {
+            inputRef.current = node; // TextField 내부의 inputRef 사용
+            if (typeof ref === 'function') ref(node); // RefCallback이 오는 경우 적용
```
이 훅을 한 번 참고해보세요 ! [useForwardRef](https://github.com/facebook/react/issues/24722)

Sangjun-man replied:
오호 보내주신 링크 읽어봤는데 useForwardRef를 사용하면, 지금 방식처럼 콜백함수를 넣지 않고 평서 useRef 쓰듯 ref로 선언한 값을 넣어주면 되겠네요!!!
수정해보겠습니다 감사합니다~~!

Paperkeem commented on src/components/TextField/Example.tsx:
```diff
@@ -0,0 +1,107 @@
+        }}
+      ></div>
+
+      <form
+        onSubmit={handleSubmit(onSubmit)}
+        style={{
+          height: '100vh',
+          display: 'flex',
+          flexDirection: 'column',
+          justifyContent: 'space-around'
+        }}
+      >
```
상준님 고생 많으셨습니다~!

제가 ImageUploder 제작할 때 useFormContext라는 API를 사용했어요. (바로 상위 부모 컴포넌트의 useForm return값을 props로 받지않고 사용할 수 있다고 합니다.)

TextField와 ImageUploder를 같이 쓰기 위해서
상위에서 useForm return 값 선언해주고 
FormProvider component로 감싸서 사용하면 동일하게 동작합니다!

```
  const methods = useForm();
  const { handleSubmit, control, register } = methods;
```

```suggestion
      <FormProvider {...methods}>
       <form
          onSubmit={handleSubmit(onSubmit)}
          style={{
            height: '100vh',
            display: 'flex',
            flexDirection: 'column',
            justifyContent: 'space-around'
          }}
        >
```

제가 FormProvider랑 form tag는 결합해서 쓸 수 있도록 custom component 한개 만들어서 추가로 pr 올릴게요!

[useFormContext 문서](https://react-hook-form.com/docs/useformcontext)

Sangjun-man replied:
오옹 useFormContext 사용하면 다른 컴포넌트들에 대해서도 Context 설정이 가능하군요!! 알겠습니다
다만 지금 validation 하는 방식이 TextField 내부적으로 진행하는 방식이라,  react hook form 내의 state까지는 반영되지 않을것 같습니다. 이부분 추가해서 커밋해서 올려보겠습니다~~

hayoiii commented on src/components/TextField/TextField.css.ts:
```diff
@@ -0,0 +1,128 @@
+
+export const input = recipe({
+  variants: {
+    size: {
+      big: variants.h3,
+      small: variants.body2
```
size variant는 관례적으로 xsmall, small, medium, large, xlarge 와 같이 사용하는 것 같아요
추후에 size variant가 추가될 수도 있으니까 big보다는 확장성이 있는 large를 사용하는 거 어떠세요?

Sangjun-man replied:
사이즈 변수명 같은 경우는 figma 디자인 해주신대로 이름지었는데, 관례적으로 사용하는게 더 확실한 방법인것 같네요! 
Input 관련 size variable이 추가 될 지에 대해서는 저 스스로 알수 없을것 같아서 디자이너님께 변수명 변경에 대해 얘기 나눠보고, 피그마에도 해당 내용 반영될 수 있게끔 커뮤니케이션 해보겠습니다!

hayoiii commented on src/components/TextField/TextField.css.ts:
```diff
@@ -0,0 +1,128 @@
+      variants: {},
+      style: {
+        vars: {
+          '--status-color': palette.gray300
```
(마이너) vanillia extract에 createVars라는 함수를 사용하면 css 변수를 하드코딩하지 않아도 됩니다
``` typescript
const statusColor = createVars()
```
```suggestion
         [statusColor]: palette.gray300
```
```typescript
export const underbar = style({
  borderColor: statusColor
});
```

Sangjun-man replied:
헉 제가 너무 원하던 기능이었어요.. 알려주셔서 감사합니다!!

hayoiii commented on src/components/TextField/TextField.css.ts:
```diff
@@ -0,0 +1,128 @@
+    visibility: 'hidden'
+  },
+  variants: {
+    status: {
+      on: {
+        visibility: 'visible'
+      },
+      off: {
+        visibility: 'hidden'
+      }
+    }
```
```suggestion
    visible: {
      true: {
        visibility: 'visible'
      },
      false: {
        visibility: 'hidden'
      }
    }
```
on이거나 off인 양자 택일 값이면 boolean을 이용하는 게 더 깔끔할 것 같아요
이렇게 바꾸면 boolean 값으로 쉽게 variant를 설정할 수 있습니다
```typescript
  icon({ visible: true })
```

Sangjun-man replied:
recipe variants에 과연 boolean 값으로 접근할수 있을까? 라는 생각만 갖고 이런 방식으로 작성했었는데,, 수정하겠습니다!!

hayoiii commented on src/components/TextField/TextField.tsx:
```diff
@@ -0,0 +1,247 @@
+  inputRef: { current: HTMLInputElement | null };
+  status: TextFieldStatus;
+}) => {
+  const checkIsOnOff = () =>
+    inputRef.current && inputRef.current.value.length && status !== 'default'
+      ? 'on'
+      : 'off';
```
위 코멘트와 이어집니다
style.icon의 variant를 boolean값으로 변경한다면,
- 삼항 연산자가 필요 없어집니다. 단순 boolean값을 반환해도 됩니다
- boolean을 사용하면 타입 체크도 간편합니다
- `checkIsOnOff` 라는 함수명도 `isOn`, `isVisible`, `isActivated`와 같이 더 단순한 형태로 바꿀 수 있습니다

```suggestion
  const isVisible = () =>
    inputRef.current && inputRef.current.value.length && status !== 'default'
```

Sangjun-man replied:
확실히 더욱 깔끔하네요!! 적용해서 커밋 올리겠습니다!

hayoiii commented on src/components/TextField/TextField.tsx:
```diff
@@ -0,0 +1,247 @@
+        className={`${max && clsx([style.count, variants.body3])}`}
+      >
+        {max && `0/${max}`}
+      </div>
```
typography 컴포넌트(`<Body3/>`) 대신 typography variants를 임포트해서 쓰시는 이유가 있으실까요??

```suggestion
      <Body3
        ref={ref}
        className={clsx({[style.count]: Boolean(max)})}
      >
        {max && `0/${max}`}
      </div>
```

hayoiii commented on src/components/TextField/TextField.tsx:
```diff
@@ -0,0 +1,247 @@
+      {status === 'loading' ? (
+        <p>loading</p>
+      ) : (
+        <p className={clsx([style.message, variants.caption2])}>{message}</p>
```
위 코멘트와 동일합니다
```suggestion
        <Caption2 className={style.message}>{message}</Caption2>
```

Sangjun-man replied:
입력창 우측에 현재 입력된 글자수 / 최대글자수를 보여주는 `<Count />` 컴포넌트에서는 React.forwardRef로  이미 만들어진 ref를 전달받다보니 만들어주신 Typograph 컴포넌트에 적용되지 않는 오류가 생겨서, 클래스네임을 import해서 사용했었습니다.

그런데 위에 @naro-Kim 님이 말씀해주신 useForwardRef() hooks로 ref를 생성하면, 해당 부분을 클래스네임이 아닌 Typography 컴포넌트로 사용할 수 있을것같아서, 이부분 수정해서 커밋 올리겠습니다~

Sangjun-man replied:
@hayoiii 적용해보니, Typography 컴포넌트 자체가 React.forwardRef로 작성되어야 외부ref 적용이 가능해서, 기존 방식대로 클래스명을 임포트해서 사용해야할것 같습니다!

hayoiii commented on src/components/TextField/hooks/useTextFieldStatus.ts:
```diff
@@ -0,0 +1,73 @@
+
+// eslint-disable-next-line no-unused-vars
+type TextFieldMessagesType = { [key in TextFieldStatus]: string };
+type TextFieldStatusDto = [TextFieldStatus, string];
```
여기서 아래처럼 오브젝트가 아니라 어레이 타입으로 선언하신 이유가 궁금해요

```typescript
type TextFieldStatusDto = {
  status: TextFieldStat;
  message: string;
}
```

위처럼 선언하면 `const message = receivedStatus[1]` 대신 `const status = receivedStatus.message` 이런식으로 사용할 수 있을 거 같은데 다른 이유가 있으신가요??

Sangjun-man replied:
어레이로 사용한 이유는 setState() 안에 들어가는 파라미터로 인해 코드가 너무 길어지는 것 같아서였습니다.
setState 파라미터로 object를 넣으면 프리티어 설정으로 인해 2~3줄이 늘어나더라구요

그리고, 리팩토링 전에는 기존에는 setState()에 해당하는 함수를 그냥 내보냈었는데,  hooks 외부에서 (ex. TextField의 handle 함수) setState 함수를 쓸때마다 2~3줄이 추가로 늘어나버려서, 가독성 측면에서 상태를 array로 바꿨었습니다.


그런데 다시 찬찬히 보니 오브젝트로 다시 바꾸는게 맞는것 같아요,
리팩토링 이후 setState를 한번 감싼 setTextFieldStatus 함수를 리턴하고 있어 파라미터를 꼭 object로 받지 않아도 되면서 위와 같은 문제가 해결이 될것 같습니다.

체크해주셔서 감사합니다~ 이부분도 수정해서 커밋 올리겠습니다!

hayoiii reviewed (commented):
복잡한 로직 짜시느라 너무 고생하셨습니다 감사합니다 👍🏻

Sangjun-man committed:
🤖 refactor: textFieldState 타입 변경, 메서드 명 수정

- textFieldState의 상태를 Array에서 Object로 수정했습니다.
- useTextFieldState hooks가 리턴하는 메서드 중 setTextFieldStatus 메서드 이름을
  updateTextFieldState로 변경했습니다.

Sangjun-man committed:
🤖 refactor: css 변수 createVar로 처리, icon recipe 변수명 수정, 일부 스타일 수정

Sangjun-man committed:
🤖 refactor: useForwardRef hook 추가
